providers/common/der/der_writer.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include "internal/cryptlib.h"
+#include "crypto/bn.h"
+#include "prov/der.h"
+
+/*
+ * All functions return the same:
+ * == 0 : error
+ * >  0 : total length so far
+ */
+
+static size_t int_der_alloc(unsigned char **pp, const unsigned char *start,
+                            size_t n)
+{
+    if (pp != NULL) {
+        if (*pp - n < start)
```

Comment:
I'm not sure I understand.  `*pp` is assumed to point inside a buffer that starts at `start`, and the check is to determine if decrementing `*pp` by `n` would cause `*pp` to end up outside that buffer.  How would you do that?

Reply:
If `start + n` is guaranteed to be within the allocated array then the expression can be rewritten as `*pp < start + n`. But even better is `*pp - start < n` as `*pp` is guaranteed to be within the buffer that starts at `start` by your definition.

Reply:
Right: `*pp - start < n` should be fine.

Reply:
Okie, will do. I have a recollection that VMS C has its intricacies regarding pointer diffs, will have to look at that, but otherwise, that formula works.

Reply:
Fixed